Soft-AP settings

katast opened the pull request:
Changes in the Makefile to add Soft-AP setting's variables, those hard-coded variables will be read on wifi init and added to Soft-AP configuration.
Server side variables values and range check, only alphanumeric chars and underscore are allowed. 

Added a web page to let user change the Soft-AP main settings.

Give it a try and let me know if you find any major issue.

Commit:
Added SOFTAP hard-coded settings

Set the values of SOFTAP config straight in the Makefile

Commit:
Soft-AP Settings

Added Soft-AP settings support both in Makefile and web interface

Commit:
Compiler error

Even if I use the same compiler those errors show up. I had to change
some bool to uint8_t to make it works

Commit:
Small changes

Commit:
New function to check string

If there are only alphanumeric chars (including underscore) return 1
else return 0

Commit:
Auth_Mode var in numbers instead of string

Commit:
Disable configWipe

Commit:
Soft-AP server side variables check

Proper check of Soft-AP setting’s variables values on server side.
Only alphanumeric values and underscore are allowed in SSID and
Password. Now you can set Soft-AP as OPEN AP with custom SSID.

tve commented on Makefile:
```diff
@@ -30,7 +59,7 @@ XTENSA_TOOLS_ROOT ?= $(abspath ../esp-open-sdk/xtensa-lx106-elf/bin)/
 # Base directory of the ESP8266 SDK package, absolute
 # Typically you'll download from Espressif's BBS, http://bbs.espressif.com/viewforum.php?f=5
-SDK_BASE	?= $(abspath ../esp_iot_sdk_v1.5.0)
+SDK_BASE	?= $(abspath ../esp-open-sdk/esp_iot_sdk_v1.5.0)
```
Please don't change the defaults.

katast replied:
Sorry, I forgot to restore the original one

tve commented on esp-link/cgiwifi.c:
```diff
@@ -48,100 +56,100 @@ static char *wifiPhy[]  = { 0, "11b", "11g", "11n" };
-  for (int i = 0; i < 4; i++) {
-    if (wifi_state_change_cb[i] != NULL) (wifi_state_change_cb[i])(wifiState);
-  }
+    switch (evt->event) {
```
Please don't change indentation. It makes it difficult to review...

katast replied:
I know, I am new to git-hub. The main problem was that the files are unreadable in Xcode, so I had to change the indentation to understand the code. I'll try visual studio..

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
-      DBG("Wifi got IP. Going into STA mode..\n");
-      wifi_set_opmode(1);
-      os_timer_arm(&resetTimer, RESET_TIMEOUT, 0); // check one more time after switching to STA-only
+            wifi_set_opmode(1);
```
Does this cause a hiccup if we're already in opmode==1? Would be safer to prefix with `if (opmode != 1)`

katast replied:
Absolutely right

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+    // Do we need a password?
+    int pass_need=1;
+    
+    // Check extra security measure
```
what is this?

katast replied:
I forgot to remove this testing purpose call to wifi_softap_get_config.

Because I did't want to create several char array to each parameters and call httpdFindArg only when needed, I used this pass_need (bool would be better) to know if I need to get password or auth mode arguments,otherwise it would save a bit of memory and cpu time.

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+        return HTTPD_CGI_DONE;
+    }
+    // Set new PASSWORD
+    if( pass_need ){
```
at this point pass_need is always ==1. What is the intent of pass_need?

katast replied:
I did't want to create several char array to store each parameter and call httpdFindArg only when needed, I used pass_need to know if I need to get further looking for password and auth mode arguments, depending on ssid value and the password value itself. 
This way if ssid or password are invalid or out of range, it would save some memory and cpu time.

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+    // Get the new SSID and set
+    len=httpdFindArg(connData->getArgs, "ap_ssid", buff, sizeof(buff));
+    if(checkString(buff) && len>7 && len<32){
+        // STRING PREPROCESSING DONE IN CLIENT SIDE
```
what preprocessing?

katast replied:
In wifiAP.js line 31, I have added this lines to avoid an ajax request, if there are invalid characters in SSID or PASSWORD

```
    var clean = inputs[i].value.replace(/[^\w]/gi, "");
    var comp = clean.localeCompare(inputs[i].value);
    if ( comp != 0 ){
        showWarning("Invalid characters in " + specials[inputs[i].name]);
        return;
    }   
```

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+    len=httpdFindArg(connData->getArgs, "ap_ssid", buff, sizeof(buff));
+    if(checkString(buff) && len>7 && len<32){
+        // STRING PREPROCESSING DONE IN CLIENT SIDE
+        memset(apconf.ssid, 0, 32);
```
Should be os_memset

katast replied:
Done

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+            }
+        }
+    }else{
+        apconf.authmode = 0;
```
If a password was specified, the default authmode should be WPA2 IMHO

katast replied:
You are right

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+        wifi_set_opmode(next_mode&3);
+        
+        if (previous_mode == 2) {
+            // movint to STA-only mode from AP, so reset into STA+AP after a timeout if we don't get an IP address
```
This if clause is reached if the previous mode was STA-only, the new mode could be AP-only or AP+STA. So I don't think the actions are correct.

katast replied:
I have two different esp-link branch so I forgot to modify the original comments. 
I think actions are coherent:
if we are going to AP, disconnect STA
if we come from AP, we are going to STA or STA+AP, so a call to STA conf is needed
if we come from STA, we are going to AP or STA+AP, so a call to Soft-AP conf is needed

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+            os_timer_arm(&resetTimer, RESET_TIMEOUT, 0);
+        }
+        if(previous_mode == 1){
+            // moving to STA or STA+AP, so softap config call needed
```
Why is a softap call needed if moving to STA-only mode?

katast replied:
Some as above

tve commented on esp-link/cgiwifi.c:
```diff
@@ -281,239 +298,427 @@ int ICACHE_FLASH_ATTR cgiWiFiScan(HttpdConnData *connData) {
+    "Switch to <a href=\\\"#\\\" onclick=\\\"changeWifiMode(1)\\\">STA mode</a>",
+};
+
+static char *apWifiWarn[] = { 0,
```
Why is a separate set of strings needed? They use up very scarce static RAM...

katast replied:
Right, I have changed this to cover all the possible options

tve commented on esp-link/cgiwifi.c:
```diff
@@ -524,114 +729,233 @@ static char *wifiWarn[] = { 0,
-  return HTTPD_CGI_DONE;
+// Check string againt invalid characters
+int ICACHE_FLASH_ATTR checkString(char *str){
+    unsigned char strStripped[50];
```
Why is a copy of the string needed? You could just return 0 when encountering the first invalid char? Maybe you changed your mind about what this function needed to do?

katast replied:
It's just because I am new to git-hub and as you said reading diff with new indentation was terrible, I used this function in an other contest. I have changed this to act as expected

tve commented on esp-link/cgiwifi.c:
```diff
@@ -524,114 +729,233 @@ static char *wifiWarn[] = { 0,
+    wifi_station_get_config_default(&stconf);
+    wifi_softap_get_config_default(&apconf);
+    
+#ifdef CGIWIFI_DBG
```
Please use DBG macro...

katast replied:
Done

tve commented on esp-link/cgiwifi.c:
```diff
@@ -524,114 +729,233 @@ static char *wifiWarn[] = { 0,
+    
+    // STATION parameters only on a full flash, because default opmode is 2
+#if defined(STA_SSID) && defined(STA_PASS)
+    if( x == 2 ){
```
Why only in 2 and not also in opmode 3?

katast replied:
Right

Commit:
Back to original config

I forgot to restore to original

Commit:
Added to avoid possible hiccup when opmode already 1

Commit:
After test it results useless, so removed

Commit:
Changed to standard os_memset(), instead of memset()

Commit:
Redundant

Added an if statement to APs scans cgi to avoid scan in AP mode.
Reduced to one set of warnings and add a fourth line to cover the only
AP mode option in Soft-AP setting´s page

Commit:
Reduced to essential

It was to much code to such a tiny task

Commit:
Changed to Macro debug

Commit:
Set Station parameters if defined

To allow changes both in STA or AP settings, they should be active.
Opmode is set to 3 few lines above, so changes allowed in both STA and
AP.
Because the above opmode set, this “if” is useless.

Commit:
Modified to cover all possible options

Case scenario:
-new AP SSID without password, keep old password and auth mode
-new AP SSID with password empty, auth mode set to OPEN
-new AP SSID with password, default auth mode WPA2

Commit:
Int to bool

Commit:
Add default WPA2 when a password was specified

Commit:
Modified comments to match actions

Commit:
Back to default

Commit:
Back to default

Commit:
Fix indentation

katast commented on esp-link/cgiwifi.c:
```diff
@@ -265,14 +274,21 @@ static int ICACHE_FLASH_ATTR cgiWiFiGetScan(HttpdConnData *connData) {
+    }else{
+        jsonHeader(connData, 404);
+        return HTTPD_CGI_DONE;
+    }
```
Avoid APs scan in AP mode, because it's not allowed

Commit:
Fix Indentation

Commit:
Fix indent

Commit:
Fix indent

katast commented on esp-link/cgiwifi.c:
```diff
@@ -300,19 +317,19 @@ static void ICACHE_FLASH_ATTR resetTimerCb(void *arg) {
 }
 
-// Temp store for new ap info.
-static struct station_config stconf;
```
Moved to the top with softap_conf

katast commented on esp-link/cgiwifi.c:
```diff
@@ -525,43 +717,50 @@ static char *wifiWarn[] = { 0,
+    int p = wifi_get_phy_mode();
+    char *phy = wifiPhy[p&3];
+    char *warn = wifiWarn[op];
+    if (op == 3) op = 4; // Done to use only one set of warnings
```
In Soft-AP settings when in STA+AP mode, let user switch to AP mode. 
To use only one set of warnings an extra line is needed, to cover all cases.

Commit:
Fix indent

Commit:
From os_printf to DBG

Commit:
Comments review
